### PR TITLE
Code Refinement: Quality Enhancements and Strategic TODOs

### PR DESCRIPTION
**Pull request by ntomita**
I've made several updates to improve code quality and prepare for future revisions:

- Packaging Refactoring: Fixed the packaging for pip.
- Code Quality: Refactored code for better readability and robustness.
- Edge Case Handling: Addressed potential edge cases.
- Config: Updated missing config values .
- TODOs: Added TODOs as markers for areas needing future team discussions and planning.

Apologies for the mix of bug fixes and packaging refactoring in one PR – it wasn't the initial plan, but that's how it unfolded. 

Also, due to recent restructuring, MaskHIT/configs/config_default.yaml's history isn't reflected in MaskHIT/maskhit/configs/config_default.yaml. For detailed changes, please perform a local diff. Apologies for any inconvenience.

Your review and feedback, especially on the TODOs for future planning, would be highly appreciated. Thanks!

### CODE REVIEW

File: maskhit/configs/config_default.yaml [hunk 1]
```diff
@@ -9,35 +13,45 @@
...
   meta_svs: !!str <SET_YOUR_OWN_VALUE>
 
   # Path to the folder containing meta info about the dataset
   # Example: ../../../SlidePreprocessing/for_vit/meta/ibd_project_meta.pickle
+  #TODO: Generated by SlidePrep/MaskHIT_Prep/01_get_svs_meta.py
   meta_all: !!str <SET_YOUR_OWN_VALUE>
 
   # Outcome our model is trying to predict
   # Example: "Dx (U=UC, C=Cr, I=Ind)"
+  #TODO: Not sure what this is meant. In what format?
```
Comment (adas2125) [inline maskhit/configs/config_default.yaml]: This parameter is a string and it should be the name of a column present in the dataset that you want the models to predict. In the example given, "Dx (U=UC, C=Cr, I=Ind)" was the name of the labels column.

File: maskhit/configs/config_default.yaml [hunk 2]
```diff
@@ -9,35 +13,45 @@
...
   study: !!str <SET_YOUR_OWN_VALUE>
 
   # type of disease; whether it is cancer or not
+  #TODO: Add doc why this is matter
   is_cancer: !!bool False
 
   # title of project/disease name
+  #TODO: what's the difference between this and 'study'
+  #This seems to be used for folder value:??
+  #        meta_svs['folder'] = config.dataset.disease
```
Comment (adas2125) [inline maskhit/configs/config_default.yaml]: I agree. This is confusing. I think we should just have the user specify study or disease instead of having two different config parameters.

File: maskhit/configs/config_default.yaml [hunk 3]
```diff
@@ -9,35 +13,45 @@
...
   # Example: "Dx (U=UC, C=Cr, I=Ind)"
+  #TODO: Not sure what this is meant. In what format?
   outcome: !!str <SET_YOUR_OWN_VALUE>
 
   # classification type
+  #TODO: Other options? survival, classification, regression
   outcome_type: !!str classification
   
   # the name of the study
+  #TODO: suggest to rename to 'study_name'
```
Comment (11tonystark) [inline maskhit/configs/config_default.yaml]: We will add those changes in next PR

File: maskhit/configs/config_default.yaml [hunk 4]
```diff
@@ -47,9 +61,11 @@ patch:
   num_patches: !!int 0
 
   # magnification level at which patches were extracted at
+  #TODO: Why manually setting this? Can we specify a config file from Prep to extract this info?
   magnification: !!int 10
 
   # intensity of weight decay
+  #TODO: Why this is under patch section?
```
Comment (adas2125) [inline maskhit/configs/config_default.yaml]: We will move this to model

File: maskhit/configs/config_default.yaml [hunk 5]
```diff
@@ -68,11 +84,13 @@ model:
   batch_size: !!int 16
 
   # which fold to use after kfold cross validation
+  #TODO: Not sure what this means. Also variable name should be more descriptive.
```
Comment (adas2125) [inline maskhit/configs/config_default.yaml]: This config argument does not seem to be used in the training script

File: maskhit/options/train_options.py [hunk 6]
```diff
@@ -67,6 +67,7 @@ def initialize(self):
                                  type=str,
                                  default='',
                                  help='select cancer subset, if empty then use entire dataset')
+        #TODO: lack of doc. so this 'fold' means 0-indexed test fold ID?
```
Comment (adas2125) [inline maskhit/options/train_options.py]: Yes. This argument refers to what fold we want to test our model on when we specify the mode as test

File: maskhit/options/train_options.py [hunk 7]
```diff
@@ -320,10 +321,12 @@ def initialize(self):
                                  help='turn off strict mode')
 
         # patch region masking
+        #TODO: Avoid too similar arg names
         self.parser.add_argument('--prob-mask',
                                  type=float,
                                  default=0,
                                  help='mask probability in BERT')
+        #TODO: Need doc for "masked:original:random". What are these?
```
Comment (adas2125) [inline maskhit/options/train_options.py]: proportion for masking tokens; used in constructor \
                                    of MaskingGenerator class, masked:original:random. Masked represents \
                                        proportion of patches of WSI not seen to the model during training \
                                            and that the model must predict
  We added this documentation in a previous PR

File: maskhit/train.py [hunk 8]
```diff
@@ -279,17 +329,23 @@ def main():
     # loading datasets
     meta_svs = pd.read_pickle(config.dataset.meta_svs) 
 
+    #TODO: Need doc why this option exists and what it does.
```
Comment (adas2125) [inline maskhit/train.py]: This seems to filter out the dataset based on which slides are FFPE and which are not. The labels should be assigned in the pre-processing step of the meta files.

File: maskhit/train.py [hunk 9]
```diff
@@ -279,17 +329,23 @@ def main():
...
         meta_all = pd.read_pickle(config.dataset.meta_all)
+        #Debug:
+        print("Debug: meta_all:\n", meta_all.head(2))
+
+        #TODO: mode=extract is not expected in the train_options. Need doc.
         if args.mode == 'extract':
             meta_train = meta_val = meta_all
         elif 'fold' in meta_all.columns:
             if meta_all.fold.nunique() == 5:
+                #TODO: Why expecting to have 5 folds? Generalize.
```
Comment (11tonystark) [inline maskhit/train.py]: We will add it to config file

File: maskhit/train.py [hunk 10]
```diff
@@ -56,13 +66,15 @@
 print(f"TEST: {config.dataset.meta_svs}")
 
 args.all_arguments = ' '.join(sys.argv[1:])
+#TODO: this var not used
```
Comment (11tonystark) [inline maskhit/train.py]: We added it to look at the dataset info but we can remove it

File: maskhit/train.py [hunk 11]
```diff
@@ -195,13 +210,38 @@ def prepare_data(meta_split, meta_file, vars_to_include=[]):
...
+        split_value = row['barcode']
+        ids_to_add.append(split_value)
+    meta_split['id_patient'] = ids_to_add
+    # Define the lambda function
+    shorten_id = lambda x: x.split('-')[0] + '-' + x.split('-')[1] + '-' + x.split('-')[2]
+    # Apply the lambda function to the id_patient column
+    meta_file['id_patient'] = meta_file['id_patient'].apply(shorten_id)
+
+
+
```
Comment (11tonystark) [inline maskhit/train.py]: Sure

File: MaskHIT/maskhit/configs/config_default.yaml [hunk 12]
```diff
@@ -96,6 +96,27 @@ model:
...
+  lr_attn: !!float 1e-5 # lowered since we are using pre-trained model
+  lr_fuse: !!float 1e-4
+  lr_loss: !!float 1e-4
+  lr_pred: !!float 7e-4
+
+  performance_measure: !!str f1
+
+  accumulation_steps: !!int 1
+  dropout: !!float 0.2
+  #TODO: ?
```
Comment (adas2125) [inline MaskHIT/maskhit/configs/config_default.yaml]: Should we separate the batch size for visualization and training into 2 different config parameters?

File: MaskHIT/maskhit/trainer/metrics.py [hunk 13]
```diff
@@ -34,6 +34,7 @@ def find_confident_instance(preds):
     return preds[preds.max(1).argmax()]
 
 def read_and_adjust_csv(file_name, last_max_id):
+    #TODO: I doubt this function is generalizable. Need review.
```
Comment (adas2125) [inline MaskHIT/maskhit/trainer/metrics.py:46]: This function is reading from the log files that were generated as a result of training the model. The columns of the log files are in the format:
Expected CSV Format:
        - The second column should contain numeric IDs
        - Columns 3 to 6 should contain data representing predictions
        - Column 7 should contain targets
    Note: Above follows the format of prediction files generated by testing.
However, I think it may be a good idea for another review. Thank you for pointing this out.

File: MaskHIT/maskhit/trainer/metrics.py [hunk 14]
```diff
@@ -95,8 +96,12 @@ def aggregate_predictions(study_name, timestr, classes, num_files = 5):
 
 
 def calculate_metrics(ids, preds, targets, label_classes, outcome_type='survival', mode = ''):
+    #TODO: Should support all the outcome_types, or should raise error if a provided type is not available.
+    #Can we create a class with the logics and feed the object, instead of implementing the logic here.
```
Comment (adas2125) [inline MaskHIT/maskhit/trainer/metrics.py:109]: Yes, I think this would be a cleaner solution

File: MaskHIT/maskhit/options/train_options.py [hunk 15]
```diff
@@ -301,7 +301,7 @@ def initialize(self):
         # experiment optional options
         self.parser.add_argument('--checkpoints-folder',
                                  type=str,
-                                 default='checkpoints_new',
```
Comment (11tonystark) [inline MaskHIT/maskhit/options/train_options.py:301]: We renamed the checkpoints to avoid the confusion.
One type were generated from pre-trained and other were generated from fine tuned
